src/front/wgsl/lower/mod.rs

```diff
@@ -317,17 +319,70 @@ impl<'source, 'temp, 'out> ExpressionContext<'source, 'temp, 'out> {
         &mut self,
         expr: crate::Expression,
         span: Span,
-    ) -> Handle<crate::Expression> {
+    ) -> Result<Handle<crate::Expression>, Error<'source>> {
         match self.expr_type {
-            ExpressionContextType::Runtime(ref mut ctx) => ctx.naga_expressions.append(expr, span),
-            ExpressionContextType::Constant => self.module.const_expressions.append(expr, span),
+            ExpressionContextType::Runtime(ref mut rctx) => {
+                let mut eval = ConstantEvaluator {
+                    types: &mut self.module.types,
+                    constants: &self.module.constants,
+                    expressions: rctx.naga_expressions,
+                    const_expressions: Some(&self.module.const_expressions),
+                    append: Some(
+                        |arena: &mut Arena<crate::Expression>, expr: crate::Expression, span| {
+                            let is_running = rctx.emitter.is_running();
+                            let needs_pre_emit = expr.needs_pre_emit();
+                            if is_running && needs_pre_emit {
+                                rctx.block.extend(rctx.emitter.finish(arena));
+                            }
+                            let h = arena.append(expr, span);
+                            if is_running && needs_pre_emit {
+                                rctx.emitter.start(arena);
+                            }
+                            h
+                        },
+                    ),
+                };
+
+                let res = eval
+                    .try_eval_and_append(&expr, span)
+                    .map_err(|e| Error::ConstantEvaluatorError(e, span));
```

Comment:
This `map_err` is meaningless because we're just going to drop the error on the floor below, right?